File: python/day05/day05.py
from enum import Enum

import numpy


class Method(str, Enum):
    ONE_BY_ONE = "one by one"
    ALL = "all"


def move_crates(
    moves: list[str], stacks: list[list[int]], method: Method
) -> str:
    for move in moves:
        amount = int(move.split(' ')[1])
        from_stack = int(move.split(' ')[3]) - 1
        to_stack = int(move.split(' ')[5]) - 1
        crates = stacks[from_stack][0:amount]
        stacks[from_stack] = stacks[from_stack][amount:]
        if method == Method.ONE_BY_ONE:
            stacks[to_stack] = list(reversed(crates)) + stacks[to_stack]
        if method == Method.ALL:
            stacks[to_stack] = crates + stacks[to_stack]
    return ''.join([stack[0] for stack in stacks])


with open('day05/input', encoding='utf-8') as f:
    lines = [line for line in f]

stacks = [
    [crate for crate in stack if crate != ' ']
    for stack in numpy.array([
        [pl[(i*4)+1] for i in range(0, 9)]
        for pl in lines[:8]
    ]).T.tolist()
]
moves = lines[10:]

print(f'Part 1: {move_crates(moves, stacks.copy(), Method.ONE_BY_ONE)}')
print(f'Part 2: {move_crates(moves, stacks.copy(), Method.ALL)}')

File: typescript/src/day05/day05.ts
import fs from "fs";

enum Method {
  OneByOne,
  All,
}

const doMoves = (
  stacks: Array<Array<string>>,
  moves: string[],
  method: Method
): string => {
  const mutableStacks = JSON.parse(JSON.stringify(stacks));
  moves.forEach((move) => {
    const amount = parseInt(move.split(" ")[1]);
    const fromStack = parseInt(move.split(" ")[3]) - 1;
    const toStack = parseInt(move.split(" ")[5]) - 1;
    const crates = mutableStacks[fromStack].slice(0, amount);
    mutableStacks[fromStack] = mutableStacks[fromStack].slice(
      amount,
      mutableStacks[fromStack].length
    );
    switch (method) {
      case Method.OneByOne: {
        mutableStacks[toStack] = [
          ...crates.reverse(),
          ...mutableStacks[toStack],
        ];
        break;
      }
      case Method.All: {
        mutableStacks[toStack] = [...crates, ...mutableStacks[toStack]];
        break;
      }
    }
  });
  return mutableStacks.map((stack) => stack[0]).join("");
};

const lines = fs
  .readFileSync("src/day05/input")
  .toString("utf-8")
  .split("\n")
  .filter((line) => line != "");

const moves = lines.slice(9);
const stacks = [...Array(9).keys()].map((stack) => {
  return [...Array(8).keys()]
    .reverse()
    .map((crate) => lines[crate][stack * 4 + 1])
    .filter((crate) => crate !== " ")
    .reverse();
});

console.log("Part 1: " + doMoves(stacks, moves, Method.OneByOne));
console.log("Part 1: " + doMoves(stacks, moves, Method.All));
